include/utilities/logging_utils.h

```diff
@@ -0,0 +1,96 @@
+#ifndef NGEN_LOGGING_UTILS_H
+#define NGEN_LOGGING_UTILS_H
+
+#include <iostream>
+#include <cstring>
+#include <string>
+
+#define MAX_STRING_SIZE 2048
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+namespace logging {
+    /**
+     * Send debug output to std::cerr
+     * @param msg The variable carries the humanly readable debug text info.
+     */
+    inline void debug(char* msg)
+    {
+        #ifndef NGEN_QUIET
+            std::string msg_str = std::string(msg);
+            if (msg_str.size() > MAX_STRING_SIZE) {
```

Comment:
The purpose of `MAX_STRING_SIZE` is not to prevent clients from giving us messages larger than we want to read. Rather, it is to give a bound to C implementations that may need to allocate specific buffer sizes, and perhaps to specify a point at which reading should be stopped if the string is not properly null-terminated. If this latter is the case, std::string::size is probably not going to work anyway.

For a C++ implementation as we have here, what probably makes the most sense is to ignore it. If we get a string that's not properly terminated, we probably can't know what to do with it anyway. Just pass the input to std::cerr... someone else might disagree, the most we could do is check the C-string length with something like `strnlen`, but this has the same problem as std::string::size. I would probably just ignore the `MAX_STRING_SIZE` in this implementation.

This simplifies these functions significantly, allowing you to get rid of multiple variables and copies of the string.